docs/developer/set-goals.md

```diff
@@ -79,11 +79,47 @@ The DSL includes support for logical operations. For example, this will build al
 whenPushSatisfies(IsMaven, not(IsKotlin)).setGoals(BuildGoals),
 ```
 
-## PushRule
-
 Each argument to [`sdm.withPushRules`](https://atomist.github.io/sdm/interfaces/_lib_api_machine_softwaredeliverymachine_.softwaredeliverymachine.html#withpushrules)
 is a PushRule, contributing goals on a commit if a condition is met. That condition is a PushTest.
 
+### Advanced Push Rules
+
+#### Computed Values
+
+You can use computed `boolean` values or the results of synchronous or
+asynchronous functions returning `boolean` in the DSL, making it
+possible to bring in any state you wish. For example:
+
+```typescript
+whenPushSatisfies(IsMaven, HasSpringBootApplicationClass,
+	deploymentsToday < 25)
+    .itMeans("Not tired of deploying Spring apps yet")
+    .setGoals(LocalDeploymentGoals),
+```
+
+#### Decision Trees
+
+You can write decision trees in push rules or other push
+mappings. These can be nested to arbitrary depth, and can use computed
+state. For example:
+
+```typescript
+let count = 0;
+const pm: PushMapping<Goals> = given<Goals>(IsNode)
+	// Compute a value we'll use later
+    .init(() => count = 0)
+    .itMeans("node")
+    .then(
+        given<Goals>(IsExpress).itMeans("express")
+            .compute(() => count++)
+            // Go into tree branch rule set
+            .then(
+                whenPushSatisfies(count > 0).itMeans("nope").setGoals(NoGoals),
+				whenPushSatisfies(TruePushTest).itMeans("yes").setGoals(HttpServiceGoals),
+            ),
+    );
+```
```

Comment:
wow
Even I don't know what is going on there

Hey @johnsonr is this something we use and want to document?